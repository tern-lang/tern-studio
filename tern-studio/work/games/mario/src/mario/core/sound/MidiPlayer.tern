
/**
 * MidiPlayer.java
 * See MidiTest2.java to easily see how to use this object
 * See MidiTest1.java for a more complex application using a meta event
 * 
 * The Java Sound API provides MIDI sound capabilities. To play MIDI music, you need two
 * objects and Sequencer. A Sequence object contains the MIDI data, and a
 * Sequencer sends a Sequence to the MIDI synthesizer.
 * 
 * Generally, playing a MIDI file looks something like this:
 * var sequence = MidiSystem.getSequence(new File(filename));
 * var sequencer = MidiSystem.getSequencer();
 * sequencer.open();
 * sequencer.setSequence(sequence);
 * sequencer.start();
 * 
 * By default, the Sequencer plays a Sequence once and then stops. To loop a 
 * Sequence, you need to be notified when the music is done playing and start over.
 * This is done with a MetaEventListener.
 * 
 */

import io.BufferedInputStream;
import io.FileInputStream;
import io.IOException;
import io.InputStream;

import sound.midi.InvalidMidiDataException;
import sound.midi.MetaEventListener;
import sound.midi.MetaMessage;
import sound.midi.MidiSystem;
import sound.midi.MidiUnavailableException;
import sound.midi.Sequence;
import sound.midi.Sequencer;

class MidiPlayer with MetaEventListener {

    // Midi meta event integer for a track ending
    public static const END_OF_TRACK_MESSAGE = 47;

    private var sequencer; // play through this sequencer
    private var loopFlag;
    private var paused;

    // Creates a new MidiPlayer object.
    // Gets the sequencer to play a sequence
    public new() {
        try {
           // MidiSystem is a lot like AudioSystem, in that it allows you to
           // get the proper materials needed to begin.
            sequencer = MidiSystem.getSequencer();
            sequencer.open();
            sequencer.addMetaEventListener(this);
        }
        catch (e) {
            sequencer = null;
        }
    }
    
    // Loads a sequence from the file system. Returns null if an error occurs.
    public getSequence(filename: String) {
        try {
            return this.getSequence(getResource(filename));
        }
        catch (e) {
            //e.printStackTrace();
            return null;
        }
    }

    // Loads a sequence from an input stream. Returns null if an error occurs.
    public getSequence(is: InputStream) {
        try {
            if (!is.markSupported()) {
                is = new BufferedInputStream(is);
            }
            // get the sequence
            var s = MidiSystem.getSequence(is);
            // close the stream and return the sequence to play!
            is.close();
            return s;
        }
        catch (e) {
            //ex.printStackTrace();
            return null;
        }
        catch (e) {
            //ex.printStackTrace();
            return null;
        }
    }
    
    private getResource(filename) {
       return SoundManager.class.getModule()
                .getContext().getManager().getInputStream(filename);
     }

    // Plays a sequence, optionally looping. This method returns
    // immediately. The sequence is not played if it is invalid.
    public play(sequence, loopFlag) {
       // must have a sequencer, sequence, and sequencer must be open
        if (sequencer != null && sequence != null && sequencer.isOpen()) {
            try {
                sequencer.setSequence(sequence);
                sequencer.start();
                sequencer.setTempoFactor(1f);
                this.loopFlag = loopFlag;
            }
            catch (e) {
                e.printStackTrace();
            }
        }
    }

    // This method is called by the sound system when a meta
    // event occurs. In this case, when the end-of-track meta
    // event is received, the sequence is restarted if
    // looping is on.
    public meta(event) {
        if (event.getType() == END_OF_TRACK_MESSAGE) {
            if (sequencer != null && sequencer.isOpen() && loopFlag) {
               // tick postion required to reset the sequencer
               sequencer.setTickPosition(0);
                sequencer.start();
            } else {
               System.exit(0);
            }
        }
    }

    // Stops the sequencer and resets its position to 0.
    public stop() {
         if (sequencer != null && sequencer.isOpen()) {
             sequencer.stop();
             sequencer.setMicrosecondPosition(0);
         }
    }

    // Closes the sequencer
    public close() {
         if (sequencer != null && sequencer.isOpen()) {
             sequencer.close();
         }
    }

    // Gets the sequencer
    public getSequencer() {
        return sequencer;
    }

    // Sets the paused state. Music may not imediately pause
    public setPaused(paused) {
        if (this.paused != paused && sequencer != null && sequencer.isOpen()) {
            this.paused = paused;
            if (paused) {
                sequencer.stop();
            }
            else {
                sequencer.start();
            }
        }
    }

    // Returns the paused state
    public isPaused() {
        return paused;
    }

}

