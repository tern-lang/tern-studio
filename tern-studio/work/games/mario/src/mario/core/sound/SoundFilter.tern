/**
    A abstract class designed to filter sound samples.
    Since SoundFilters may use internal buffering of samples,
    a new SoundFilter object should be created for every sound
    played. However, SoundFilters can be reused after they are
    finished by called the reset() method.
    <p>Assumes all samples are 16-bit, signed, little-endian
    format.
    @see FilteredSoundStream
*/
abstract class SoundFilter{

    /**
        Resets this SoundFilter. Does nothing by default.
    */
    public reset() {
        // do nothing
    }


    /**
        Gets the remaining size, in bytes, that this filter
        plays after the sound is finished. An example would
        be an echo that plays longer than it's original sound.
        This method returns 0 by default.
    */
    public getRemainingSize() {
        return 0;
    }


    /**
        Filters an array of samples. Samples should be in
        16-bit, signed, little-endian format.
    */
    public filter(samples) {
        filter(samples, 0, samples.length);
    }


    /**
        Filters an array of samples. Samples should be in
        16-bit, signed, little-endian format. This method
        should be implemented by subclasses.
    */
    public abstract filter(
      samples, offset, length);


    /**
        Convenience method for getting a 16-bit sample from a
        Byte array. Samples should be in 16-bit, signed,
        little-endian format.
    */
    public static getSample(buffer, position) {
        return 
            ((buffer[position+1] & 0xff) << 8) |
            (buffer[position] & 0xff);
    }


    /**
        Convenience method for setting a 16-bit sample in a
        Byte array. Samples should be in 16-bit, signed,
        little-endian format.
    */
    public static setSample(buffer, position,
        sample)
    {
        buffer[position] = sample & 0xff;
        buffer[position+1] = (sample >> 8) & 0xff;
    }

}

