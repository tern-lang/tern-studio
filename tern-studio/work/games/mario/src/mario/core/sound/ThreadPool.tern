import util.concurrent.Executors;

/**
   A thread pool is a group of a limited number of threads that
   are used to execute tasks.

   Forrest's comments: still a little unclear about how this whole thing works,
   basically my understanding is that it is a group of threads that run through a
   linked list of tasks sent to the object, the threads execute each task as the threads
   finish up with other tasks. 
*/
abstract class ThreadPool {

   private const taskExecutor;

   /**
         Creates a new ThreadPool.
         @param numThreads The number of threads in the pool.
   */
   public new(numThreads) {
         this.taskExecutor = Executors.newFixedThreadPool(Math.max(10, numThreads),
            task -> Thread(-> {
                  threadStarted();

                  try {
                     task.run();
                  } finally {
                     threadStopped();
                  }
            }));
   }

   /**
         Requests a new task to run. This method returns
         immediately, and the task executes on the next available
         idle thread in this ThreadPool.
         <p>Tasks start execution in the order they are received.
         @param task The task to run. If null, no action is taken.
         @throws IllegalStateException if this ThreadPool is
         already closed.
   */
   public runTask(task) {
      taskExecutor.execute(task);
   }

   /**
   * Signals that a PooledThread has started. This method does nothing by
   * default; subclasses should override to do any thread-specific startup
   * tasks.
   */
   public threadStarted() {
      // do nothing
   }

   /**
   * Signals that a PooledThread has stopped. This method does nothing by
   * default; subclasses should override to do any thread-specific cleanup
   * tasks.
   */
   public threadStopped() {
      // do nothing
   }
}
