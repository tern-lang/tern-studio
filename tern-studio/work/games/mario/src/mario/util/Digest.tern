import security.MessageDigest;

class Digest {

   static digest(stream: InputStream){
      var buffer = new ByteArrayOutputStream();
      var count = 0;
      
      while((count = stream.read()) != -1) {
         buffer.write(count);
      }
      return digest(buffer.toByteArray());
   }

   static digest(buffer: Byte[]){
      return digest(buffer, 0, buffer.length);
   }

   static digest(buffer, off, len) {
      var md = MessageDigest.getInstance("MD5");
      md.update(buffer, off, len);

      var bytes = md.digest();
      var builder = new StringBuffer();

      for (var i = 0; i < bytes.length; i++) {
          builder.append(Integer.toString((bytes[i] & 0xff) + 0x100, 16).substring(1));
      }
      return builder.toString(); 
   }   
}
