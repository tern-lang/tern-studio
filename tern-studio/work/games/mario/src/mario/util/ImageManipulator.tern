import static awt.Image.*;

import util.concurrent.ConcurrentHashMap;
import awt.Graphics;
import awt.Graphics2D;
import awt.Image;
import awt.image.BufferedImage;
import io.File;
import io.IOException;

import imageio.ImageIO;

/**
 * ImageManipulator.java
 * @author Forrest
 * 
 * Provides a set of methods used to modify a BufferedImage.
 * 
 */
module ImageManipulator {

   const cache = new ConcurrentHashMap();
   
   /** Reads in a BufferedImage using the standard ImageIO.read() */
   public loadImage(filename) {
      var img = cache.get(filename);
      try {
         if(img == null) {
            img = ImageIO.read(loadResource(filename));
            cache.put(filename, img);
          }
      } catch (e) { 
         e.printStackTrace();
      }
      return img;
   } 
   
   private loadResource(filename) {
      return ImageManipulator
               .getContext().getManager().getInputStream(filename);
    }
   
   public scaleImage(img, factor) {
      if(factor.compareTo(1) != 0) {
         var w = img.getWidth();   
         var h = img.getHeight();  
         var scaled = img.getScaledInstance(-1, h * factor, 4 /*BufferedImage.SCALE_SMOOTH*/);
         var dimg = new BufferedImage(w * factor, h * factor, img.getColorModel().getTransparency());    
         var g = dimg.getGraphics();
   
         g.drawImage(scaled, 0, 0, null);
         g.dispose();
         return dimg;
      }
      return img;
      
   }

   /** Horizontally flips img. */
   public horizontalFlip(img) {   
        var w = img.getWidth();   
        var h = img.getHeight();   
        var dimg = new BufferedImage(w, h, img.getColorModel().getTransparency());     
        var g = dimg.createGraphics();   
        g.drawImage(img, 0, 0, w, h, w, 0, 0, h, null);   
        g.dispose();   
        return dimg;   
    }  
   
   /** Vertically flips img. */
   public verticalFlip(img) {   
        var w = img.getWidth();   
        var h = img.getHeight();   
        var dimg = new BufferedImage(w, h, img.getColorModel().getTransparency());   
        var g = dimg.createGraphics();   
        g.drawImage(img, 0, 0, w, h, 0, h, w, 0, null);   
        g.dispose();   
        return dimg;   
    }  

}

