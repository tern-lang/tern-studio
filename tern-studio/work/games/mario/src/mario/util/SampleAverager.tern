

class SampleAverager {
   let  count;
   let samples;
   let index;
   
   new(count){
      this.count = count;
      this.samples = [];
      this.index = 0;
   }
   
   public average() {
      let total = 0.0;
      
      for(value in samples) {
         total += value;
      }
      let average: Number = total / samples.length;
      return average.round(3);
   }
   
   public sample(value: Number) {
      if(samples.length < count){
         samples.add(value);
      } else {
         samples.set(index++, value);
         
         if(index >= samples.length) {
            index = 0;
         }
      }
   }
}