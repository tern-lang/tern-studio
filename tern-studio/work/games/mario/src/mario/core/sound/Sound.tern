import mario.util.Digest;

/**
    The Sound class is a container for sound samples. The sound
    samples are format-agnostic and are stored as a Byte array.
*/
class Sound {

    private const filename;
    private const samples;

    /**
        Create a new Sound object with the specified Byte array.
        The array is not copied.
    */
    public new(filename, samples) {
        this.filename = filename;
        this.samples = samples;
    }


    /**
        Returns this Sound's objects samples as a Byte array.
    */
    public getSamples() {
        return samples;
    }

}

