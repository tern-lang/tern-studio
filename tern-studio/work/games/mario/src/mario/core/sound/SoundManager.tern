import io.BufferedInputStream;
import io.ByteArrayInputStream;
import io.DataInputStream;
import io.FileInputStream;
import io.IOException;
import io.InputStream;

import sound.sampled.AudioFormat;
import sound.sampled.AudioInputStream;
import sound.sampled.AudioSystem;
import sound.sampled.DataLine;
import sound.sampled.LineUnavailableException;
import sound.sampled.Mixer;
import sound.sampled.SourceDataLine;
import sound.sampled.UnsupportedAudioFileException;

import mario.util.Digest;

/**
    The SoundManager class manages sound playback. The
    SoundManager is a ThreadPool, with each thread playing back
    one sound at a time. This allows the SoundManager to
    easily limit the number of simultaneous sounds being played.
    
    <p>Possible ideas to extend this class:<ul>
    <li>add a setMasterVolume() method, which uses Controls to
        set the volume for each line.
    <li>don't play a sound if more than, say, 500ms has passed
        since the request to play
    </ul>
*/

class SoundManager extends ThreadPool {

    private var playbackFormat;
    private var localLine;
    private var localBuffer;
    private var pausedLock;
    private var paused = false;

    /**
        Creates a new SoundManager using the maximum number of
        simultaneous sounds.
    */
    public new(playbackFormat):
        this(playbackFormat, getMaxSimultaneousSounds(playbackFormat)) {
    }

    /**
        Creates a new SoundManager with the specified maximum
        number of simultaneous sounds.
    */
    
    public new(playbackFormat, maxSimultaneousSounds): 
        super(Math.min(maxSimultaneousSounds, getMaxSimultaneousSounds(playbackFormat))) 
   {
        this.playbackFormat = playbackFormat;
        localLine = new ThreadLocal();
        localBuffer = new ThreadLocal();
        pausedLock = new Object();
        
        // notify threads in pool it's ok to start
        synchronized (this) {
            notifyAll();
        }
    }

    /**
        Gets the maximum number of simultaneous sounds with the
        specified AudioFormat that the default mixer can play.
    */
    public static getMaxSimultaneousSounds(playbackFormat) {
        var lineInfo = new DataLine.Info( SourceDataLine.class, playbackFormat);
        var mixer = AudioSystem.getMixer(null);
        var maxLines = mixer.getMaxLines(lineInfo);
        if (maxLines == AudioSystem.NOT_SPECIFIED) {
            maxLines = 10; // 32
        }
        return maxLines;

    }

    /**
        Does any clean up before closing.
    */
    public cleanUp() {
        // signal to unpause
        setPaused(false);

        // close the mixer (stops any running sounds)
        var mixer = AudioSystem.getMixer(null);
        if (mixer.isOpen()) {
            mixer.close();
        }
    }


    public close() {
        cleanUp();
        super.close();
    }

    public join() {
        cleanUp();
        super.join();
    }

    /**
        Sets the paused state. Sounds may not pause immediately.
    */
    public setPaused(paused) {
        if (this.paused != paused) {
            synchronized (pausedLock) {
                this.paused = paused;
                if (!paused) {
                    // restart sounds
                    pausedLock.notifyAll();
                }
            }
        }
    }

    /**
        Returns the paused state.
    */
    public isPaused() {
        return paused;
    }

    /**
        Loads a Sound from the file system. Returns null if an
        error occurs.
    */
    public getSound(filename) {
        var audioStream = getAudioInputStream(filename);
    
        if (audioStream == null) {
            return null;
        }
        // get the number of bytes to read
        var length = audioStream.getFrameLength() * audioStream.getFormat().getFrameSize();

        // read the entire stream
        var samples = new Byte[length];
        var is = new DataInputStream(audioStream);
        try {
            is.readFully(samples);
            is.close();
        }
        catch (e) {
            e.printStackTrace();
        }
        // return the samples
        return new Sound(filename, samples);
    }

    /**
        Creates an AudioInputStream from a sound from the file
        system.
    */
    public getAudioInputStream(filename: String) {
        try {
            return getAudioInputStream(getResource(filename));
        }
        catch (e) {
            e.printStackTrace();
            return null;
        }
    }

    /**
        Creates an AudioInputStream from a sound from an input
        stream
    */
    public getAudioInputStream(is: InputStream) {

        try {
            if (!is.markSupported()) {
                is = new BufferedInputStream(is);
            }
            // open the source stream
            var source = AudioSystem.getAudioInputStream(is);

            // convert to playback format
            return AudioSystem.getAudioInputStream(playbackFormat, source);
        }
        catch (e) {
            e.printStackTrace();
        }
        catch (e) {
            e.printStackTrace();
        }
        catch (e) {
            e.printStackTrace();
        }

        return null;
    }
    
    private getResource(filename) {
      return SoundManager.class.getModule()
               .getContext().getManager().getInputStream(filename);
    }

    /**
        Plays a sound. This method returns immediately.
    */
    public play(sound: Sound) {
        return play(sound, null, false);
    }

    /**
        Plays a sound with an optional SoundFilter, and optionally
        looping. This method returns immediately.
    */
    public play(sound: Sound, filter, loopFlag) {
        var is;
        if (sound != null) {
            if (loopFlag) {
                is = new LoopingByteInputStream(sound.getSamples());
            }
            else {
                is = new ByteArrayInputStream(sound.getSamples());
            }

            return play(is, filter);
        }
        return null;
    }

    /**
        Plays a sound from an InputStream. This method
        returns immediately.
    */
    public play(is: InputStream) {
        return play(is, null);
    }

    /**
        Plays a sound from an InputStream with an optional
        sound filter. This method returns immediately.
    */
    public play(is: InputStream, filter) {
        if (is != null) {
            if (filter != null) {
                is = new FilteredSoundStream(is, filter);
            }
            runTask(new SoundPlayer(this, is));
        }
        return is;
    }

    /**
        Signals that a PooledThread has started. Creates the
        Thread's line and buffer.
    */
    public threadStarted() {
        // wait for the SoundManager constructor to finish
        synchronized (this) {
            try {
                wait();
            }
            catch (e) { }
        }

        // use a short, 100ms (1/10th sec) buffer for filters that
        // change in real-time
        var bufferSize = playbackFormat.getFrameSize() * Math.round(playbackFormat.getSampleRate() / 10);

        // create, open, and start the line
        var line;
        var lineInfo = new DataLine.Info(SourceDataLine.class, playbackFormat);

        try {
            line = AudioSystem.getLine(lineInfo);
            line.open(playbackFormat, bufferSize);
        }
        catch (e) {
            // the line is unavailable - signal to end this thread
            Thread.currentThread().interrupt();
            return;
        }

        line.start();

        // create the buffer
        var buffer = new Byte[bufferSize];

        // set this thread's locals
        localLine.set(line);
        localBuffer.set(buffer);
    }

    /**
        Signals that a PooledThread has stopped. Drains and
        closes the Thread's Line.
    */
    public threadStopped() {
        var line = localLine.get();
        if (line != null) {
            line.drain();
            line.close();
        }
    }

    /**
        The SoundPlayer class is a task for the PooledThreads to
        run. It receives the threads's Line and Byte buffer from
        the ThreadLocal variables and plays a sound from an
        InputStream.
        <p>This class only works when called from a PooledThread.
    */
    class SoundPlayer with Runnable {

        private var manager;
        private var source;

        public new(manager, source) {
            this.source = source;
            this.manager = manager;
        }

        public run() {
            // get line and buffer from ThreadLocals
            var line = manager.localLine.get();
            var buffer = manager.localBuffer.get();
            if (line == null || buffer == null) {
                // the line is unavailable
                return;
            }

            // copy data to the line
            try {
                var numBytesRead = 0;
                while (numBytesRead != -1) {
                    // if paused, wait until unpaused
                    synchronized (manager.pausedLock) {
                        if (manager.paused) {
                            try {
                                manager.pausedLock.wait();
                            }
                            catch (e) {
                                return;
                            }
                        }
                    }
                    // copy data
                    numBytesRead =  source.read(buffer, 0, buffer.length);
                    if (numBytesRead != -1) {
                        line.write(buffer, 0, numBytesRead);
                    }
                }
            }
            catch (e) {
                e.printStackTrace();
            }

        }
    }
}

