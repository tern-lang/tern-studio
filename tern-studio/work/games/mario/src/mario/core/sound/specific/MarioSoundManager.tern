import io.File;
import mario.core.sound.SoundManager;
import sound.sampled.AudioFormat;
import util.Scanner;

class MarioSoundManager {

   private static const let f1 = new AudioFormat(22050, 8, 1, true, true); // 22050 Hz.
   private static const let f2 = new AudioFormat(10512, 8, 1, true, true); // 10512 Hz.
   private static const let fileName = "sounds/mario_sounds.txt";

   private let s1;
   private let s2;

   class SoundRecord {
      public let name;
      public let location;
      public let hz;
      public SoundRecord(name, location, hz) {
         this.name = name;
         this.location = location;
         this.hz = hz;
      }
   }

   public new() {
      s1 = new SoundManager(f1);
      s2 = new SoundManager(f2);

      let s = null;
      try {
         s = new Scanner(new File(fileName)); 
      } catch (e) {
         System.out.println(e);
      }

      while(s.hasNextLine()) {

         let line = s.nextLine();
         let ls = new Scanner(line);
         let sr = new SoundRecord(ls.next(), ls.next(), ls.nextInt());

      }

   }

}
