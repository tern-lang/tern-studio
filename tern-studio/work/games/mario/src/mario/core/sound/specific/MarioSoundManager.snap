import io.File;
import io.FileNotFoundException;
import util.Scanner;

import sound.sampled.AudioFormat;

import mario.core.sound.SoundManager;

class MarioSoundManager {
   
   private static const var f1 = new AudioFormat(22050, 8, 1, true, true); // 22050 Hz.
   private static const var f2 = new AudioFormat(10512, 8, 1, true, true); // 10512 Hz.
   private static const var fileName = "sounds/mario_sounds.txt";
   
   private var s1;
   private var s2;
   
   class SoundRecord {
      public var name;
      public var location;
      public var hz;
      public SoundRecord(name, location, hz) {
         this.name = name;
         this.location = location;
         this.hz = hz;
      }
   }
   
   public new() {
      s1 = new SoundManager(f1);
      s2 = new SoundManager(f2);
      
      var s = null;
      try {
         s = new Scanner(new File(fileName)); 
      } catch (e) {
         System.out.println(e);
      }
      
      while(s.hasNextLine()) {
         
         var line = s.nextLine();
         var ls = new Scanner(line);
         var sr = new SoundRecord(ls.next(), ls.next(), ls.nextInt());

      }
      
      
   }

}

